File: app/models/device.py
import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, String, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class DeviceStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    MAINTENANCE = "maintenance"
    UNKNOWN = "unknown"


class DeviceType(str, enum.Enum):
    ROUTER = "router"
    SWITCH = "switch"
    FIREWALL = "firewall"
    ACCESS_POINT = "access_point"
    LOAD_BALANCER = "load_balancer"
    OTHER = "other"


class Device(Base):
    """Network device inventory record.

    Only connection metadata is stored here. Device credentials/passwords
    are never persisted in plaintext in this table; a later phase will
    integrate a secrets vault for credential retrieval.
    """

    __tablename__ = "devices"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    hostname: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    ip_address: Mapped[str] = mapped_column(String(45), nullable=False, unique=True, index=True)
    vendor: Mapped[str] = mapped_column(String(50), nullable=False)
    device_type: Mapped[DeviceType] = mapped_column(
        Enum(DeviceType, name="device_type_enum"), nullable=False
    )
    username: Mapped[str] = mapped_column(String(100), nullable=False)
    status: Mapped[DeviceStatus] = mapped_column(
        Enum(DeviceStatus, name="device_status_enum"),
        nullable=False,
        default=DeviceStatus.UNKNOWN,
    )
    description: Mapped[str | None] = mapped_column(String(255), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: frontend/src/types/device.ts
export type DeviceStatus = "active" | "inactive" | "maintenance" | "unknown";

export type DeviceType = "router" | "switch" | "firewall" | "access_point" | "load_balancer" | "other";

export interface Device {
  id: number;
  hostname: string;
  ip_address: string;
  vendor: string;
  device_type: DeviceType;
  username: string;
  status: DeviceStatus;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface DeviceCreate {
  hostname: string;
  ip_address: string;
  vendor: string;
  device_type: DeviceType;
  username: string;
  status?: DeviceStatus;
  description?: string | null;
}

export type DeviceUpdate = Partial<DeviceCreate>;

export const DEVICE_TYPES: DeviceType[] = [
  "router",
  "switch",
  "firewall",
  "access_point",
  "load_balancer",
  "other",
];

export const DEVICE_STATUSES: DeviceStatus[] = ["active", "inactive", "maintenance", "unknown"];
